### ISSUE TITLE
regression? JsonLdError: <unprintable JsonLdError object>

### ISSUE THREAD

**Issue by yarikoptic** [digitalbazaar/pyld#78]
Our @datalad testing build bots started to puke a new error which we haven't seen before:
```
======================================================================
ERROR: datalad.metadata.tests.test_search.test_our_metadataset_search
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 133, in run
    self.runTest(result)
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 151, in runTest
    test(result)
  File "/usr/lib/python2.7/unittest/case.py", line 393, in __call__
    return self.run(*args, **kwds)
  File "/usr/lib/python2.7/unittest/case.py", line 353, in run
    result.addError(self, sys.exc_info())
  File "/usr/lib/python2.7/dist-packages/nose/proxy.py", line 128, in addError
    formatted = plugins.formatError(self.test, err)
  File "/usr/lib/python2.7/dist-packages/nose/plugins/manager.py", line 99, in __call__
    return self.call(*arg, **kw)
  File "/usr/lib/python2.7/dist-packages/nose/plugins/manager.py", line 141, in chain
    result = meth(*arg, **kw)
  File "/usr/lib/python2.7/dist-packages/nose/plugins/logcapture.py", line 237, in formatError
    return (ec, self.addCaptureToErr(ev, records), tb)
  File "/usr/lib/python2.7/dist-packages/nose/plugins/logcapture.py", line 244, in addCaptureToErr
    records + \
  File "/usr/lib/python2.7/dist-packages/nose/util.py", line 652, in safe_str
    return str(val)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 4923, in __str__
    rval += '\nCause: ' + str(self.cause)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 4923, in __str__
    rval += '\nCause: ' + str(self.cause)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 4923, in __str__
    rval += '\nCause: ' + str(self.cause)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 4917, in __str__
    rval += '\nType: ' + self.type
TypeError: cannot concatenate 'str' and 'dict' objects
```
which when I went into that environment to troubleshoot showed us another face:
```
======================================================================
ERROR: datalad.metadata.tests.test_search.test_our_metadataset_search
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/usr/lib/python2.7/dist-packages/nose/case.py", line 197, in runTest
    self.test(*self.arg)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/tests/utils.py", line 840, in newfunc
    return func(*args, **kwargs)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/tests/utils.py", line 627, in newfunc
    return t(*(arg + (filename,)), **kw)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/metadata/tests/test_search.py", line 123, in test_our_metadataset_search
    assert list(ds.search('.', report='*', regex=True))
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/metadata/search.py", line 219, in __call__
    meta = flatten_metadata_graph(meta)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/metadata/__init__.py", line 285, in flatten_metadata_graph
    return jsonld.flatten(obj, ctx={"@context": "http://schema.org/"})
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 178, in flatten
    return JsonLdProcessor().flatten(input_, ctx, options)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/venv-ci/local/lib/python2.7/site-packages/pyld/jsonld.py", line 885, in flatten
    'jsonld.FlattenError', cause=cause)
JsonLdError: <unprintable JsonLdError object>
```
yet to figure out what is going on, but I feel that it relates to having recent 1.0.{0,1} releases of pyld out ;)
attn: @mih

**Comment by yarikoptic** [comment]
ok, identified the `str` gotcha... will send a PR for that one

**Comment by yarikoptic** [comment]
With the fix in #79 it now looks more informative:
```
  File "/tmp/pyld/lib/pyld/jsonld.py", line 886, in flatten
    'jsonld.FlattenError', cause=cause)
JsonLdError: ('Could not expand input before flattening.',)
Type: jsonld.FlattenError
Cause: ('Could not perform JSON-LD expansion.',)
Type: jsonld.ExpandError
Cause: ('Dereferencing a URL did not result in a valid JSON-LD context.',)
Type: jsonld.ContextUrlError
Code: loading remote context failed
Details: {'url': u'http://schema.org/'}
Cause: ('No default document loader configured',)
Type: jsonld.LoadDocumentError
Code: no default document loader
Details: {'url': u'http://schema.org/'}  File "/tmp/pyld/lib/pyld/jsonld.py", line 4754, in _retrieve_context_urls
    remote_doc = load_document(url)
  File "/home/buildbot/datalad-pr-docker-dl-nd90/build/datalad/metadata/__init__.py", line 270, in _cached_load_document
    url, name='schema', fetcher=load_document, maxage=1

```
so it seems to be indeed a change in the interface requiring something (loader) which we didn't provide before? (would need to check now our code... later)

**Comment by davidlehn** [comment]
The changelog and readme have some info on the new doc loader situation.  There is a new aiohttp based loader along with the requests based one.  pyld now requires you to have one of those dependencies installed to get a default loader.  Not ideal to not have a default but also avoids forcing people to install dependencies they don't need.  Suggestions of a better way to handle this would be welcome.

**Comment by yarikoptic** [comment]
ok, with https://github.com/digitalbazaar/pyld/pull/80 our test seems to  pass

**Comment by davidlehn** [comment]
Released 1.0.3 with the fixes.

**Comment by yarikoptic** [comment]
Thank you for the quick release!
